src/metadata/md_provider.rs

```diff
@@ -0,0 +1,5 @@
+use crate::metadata::{MdId, Metadata};
+
+pub trait MdProvider {
+    fn retrieve_metadata(&self, md_id: &Box<dyn MdId>) -> Box<dyn Metadata>;
```

Comment:
Retrieving metedata may be failed, so here need to return a result

Reply:
Modified to return Option